stdlib/asm/math/fri.masm

```diff
@@ -0,0 +1,233 @@
+# procedure computes an iteration of fri verification for a query.
+# TODO: Check where/if some checks are needed in the beginning, like p < d (p&d are u32) and that 2^t_d == d 
+export.verify_query_layer
+    u32unchecked_div.2          #[d/2,p,C,t_d,...]
+    dup                         #[d/2,d/2,p,C,t_d,...]
+    dup.2 swap                  #[d/2,p,d/2,p,C,t_d,...]
+    u32checked_divmod           #[p%d/2,?,d/2,p,C,t_d,...]
+    swapw                       #[C,p%d/2,?,d/2,p,t_d,...]
+    dup.4                       #[p%d/2,C,p%d/2,?,d/2,p,t_d,...]
+    dup.9                       #[t_d,p%d/2,C,p%d/2,?,d/2,p,t_d,...]
+    mtree_get  swapw            #[C,V',p%d/2,?,d/2,p,t_d,...]  V' are the digest of values V = (v3,v2,v1,v0)
+    
+    # Unhash query values
+    adv_loadw                   #[V,V',p%d/2,?,d/2,p,t_d,...]   TODO: replace with adv.keyval
+    dupw movdnw.2    
+    push.4.0.0.0
+    swapw
+    push.0.0.0.0
+    rpperm
+    dropw swapw dropw
+    eqw 
+    drop dropw dropw
+
+    dupw  movdnw.2              #[v3,v2,v1,v0,p%d/2,?,d/2,p,V,t_d,...]
+    movup.2 swap                #[v3,v1,v2,v0,p%d/2,?,d/2,p,V,t_d,...]
+    dup.5                       #[?,v3,v1,v2,v0,p%d/2,?,d/2,p,V,t_d,...]
+    cdrop                       #[v1,v2,v0,p%d/2,?,d/2,p,V,t_d,...]
+    movdn.2                     #[v2,v0,v1,p%d/2,?,d/2,p,V,t_d,...]
+    dup.4                       #[?,v2,v0,v1,p%d/2,?,d/2,p,V,t_d,...]
+    cdrop   swap                #[v1,v0,p%d/2,?,d/2,p,V,t_d,...]
+                                #[v1,v0,p%d/2,?,d/2,p,V,t_d,e1,e0..]
+    dup                         #[v1,v1,v0,p%d/2,?,d/2,p,V,t_d,e1,e0..]
+    movup.12                    #[e1,v1,v1,v0,p%d/2,?,d/2,p,t_d,e0..]
+    assert_eq                   #[v1,v0,p%d/2,?,d/2,p,V,t_d,e0..]
+    dup.1                       #[v0,v1,v0,p%d/2,?,d/2,p,V,t_d,e0..]
+    movup.12                    #[e0,v0,v1,v0,p%d/2,?,d/2,p,V,t_d..]
+    assert_eq                   #[v1,v0,p%d/2,?,d/2,p,V,t_d,poe,..] #poe is position in the exponent
+    drop drop swapw             #[V,p%d/2,?,d/2,p,t_d,poe,..]
+    movup.9                     #[poe,V,p%d/2,?,d/2,p,t_d,..]
+    dup                         #[poe,poe,V,p%d/2,?,d/2,p,t_d,..]
+    movup.7                     #[?,poe,poe,V,p%d/2,d/2,p,t_d,..]
+    exec.next_pos_exp           #[poe_sq,xs,V,p%d/2,d/2,p,t_d,..]
+    swap.8 drop                 #[xs,V,p%d/2,d/2,poe_sq,t_d,..]
+    push.0                      #[0,xs,V,p%d/2,d/2,poe_sq,t_d,..]
+                                #[0,xs,V,p%d/2,d/2,poe_sq,t_d,a1,a0,..]
+    movup.11 movup.11           #[a1,a0,0,xs,v3,v2,v1,v0,p%d/2,d/2,poe_sq,t_d,..]   
+    swapw                       #[v3,v2,v1,v0,a1,a0,0,xs,p%d/2,d/2,poe_sq,t_d,..] #Should simplify fri since d1 is always equal to zero
+    movup.3 movup.3
+    exec.fold_2                 #[e1,e0,p%d/2,d/2,poe_sq,t_d,..]
+    movup.5 u32wrapping_sub.1   #[t_d-1,e1,e0,p%d/2,d/2,poe_sq,add',..]
+
+end
+
+export.outer.1
```

Comment:
Does this procedure verify the full FRI proof? If so, we should probably name it `verify`. Also, we should add comments explaining what are the expected input parameters for this procedure.

We should also state all assumptions fully here. For example, we assume degree 2 field extension and folding factor of 2.

Reply:
I think the `verify` procedure should not be calling `preprocess` internally. Instead, it should provide a well-defined interface for what it expects to receive via the stack. Then, for test purposes, we could do something like this:
```
begin
    exec.fri::preprocess
    exec.fri::verify
end
```
And maybe we will eventually move the `preprocess` procedure out of FRI module (this procedure depends on the context in which we use FRI - e.g., it could look very different when used in the context of full STARK proof verification).